### PR TITLE
LoadedPrograms cache implementation and tests

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Need facility to cache programs across active forks based on their deployment/updates.

#### Summary of Changes
This PR adds a new cache `LoadedPrograms`. The cache maintains multiple versions of loaded programs based on the active forks, and program's deployment/update slot. When the root slot is updated, the unreachable forks can be pruned from the cache.

This PR adds the framework for cache, and associated unit tests. More changes are needed in bank, accounts and surrounding code to integrate it in the flow.

Fixes #

### CODE REVIEW

File: program-runtime/src/loaded_programs.rs [hunk 1]
```diff
@@ -4,31 +4,29 @@ use {
...
     },
 };
 
 /// Relationship between two fork IDs
 pub enum BlockRelation {
-    /// The fork is an ancestor of the other fork
+    /// The slot is on the same fork and is an ancestor of the other slot
     Ancestor,
-    /// The two fork IDs are same or point to the same fork
+    /// The two slots are same and are on the same fork
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:24]: "are same" => "are equal"?

File: program-runtime/src/loaded_programs.rs [hunk 2]
```diff
@@ -4,31 +4,29 @@ use {
...
     Equal,
-    /// The fork is a descendant of the other fork
+    /// The slot is on the same fork and is a descendant of the other slot
     Descendant,
-    /// These are two parallel forks, which only have a common ancestor
+    /// The slots are on two different forks and may have had a common ancestor at some point
     Unrelated,
+    /// Either or both of the slots are either older than the latest root, or are in future
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:30]: "Either or both" => "Either one or both"?

File: program-runtime/src/loaded_programs.rs [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+        &self,
+        working_slot: &S,
+        keys: impl Iterator<Item = Pubkey>,
+    ) -> (HashMap<Pubkey, Arc<LoadedProgram>>, Vec<Pubkey>) {
+        let mut missing = Vec::new();
+        let found = keys
+            .filter_map(|key| {
+                if let Some(second_level) = self.entries.get(&key) {
+                    for entry in second_level.iter().rev() {
+                        if working_slot.is_ancestor(entry.deployment_slot) {
```
Comment (Lichtso) [inline program-runtime/src/loaded_programs.rs:196]: Also needs to check if `entry.effective_slot` is visible yet.

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -1119,6 +1121,8 @@ pub struct Bank {
     pub fee_structure: FeeStructure,
 
     pub incremental_snapshot_persistence: Option<BankIncrementalSnapshotPersistence>,
+
+    pub loaded_programs_cache: Arc<RwLock<LoadedPrograms>>,
```
[the remark this answers is not in the document]
Comment (pgarg66) [reply runtime/src/bank.rs:1125]: It's a single instance of cache shared across all the banks. The instance is created when the first bank is created on node start up. All other banks clone the Arc from their parent.

File: program-runtime/src/loaded_programs.rs (hunk 3 above)
Comment (pgarg66) [reply program-runtime/src/loaded_programs.rs]: Updated the code, and added some tests.

**Review by Lichtso — APPROVED** [review]

**Pull request merged**